examples/passport-oauth2-login/src/authentication-strategies/oauth2-strategy.ts

```diff
@@ -0,0 +1,146 @@
+// Copyright IBM Corp. 2019. All Rights Reserved.
+// Node module: example-passport-oauth2-login
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {
+  asAuthStrategy,
+  AuthenticationStrategy,
+  UserProfileFactory,
+} from '@loopback/authentication';
+import {StrategyAdapter} from '@loopback/authentication-passport';
+import {Profile} from 'passport';
+import {Strategy, StrategyOption} from 'passport-facebook';
+import {Strategy as GoogleStrategy, StrategyOptions as GoogleOptions} from 'passport-google-oauth2';
+import {bind, inject} from '@loopback/context';
+import {Request, HttpErrors} from '@loopback/rest';
+import {UserProfile, securityId} from '@loopback/security';
+import {UserService, UserServiceBindings, UserWithToken} from '../services';
+let oauth2Providers = require('../../oauth2-providers');
+
+
+const facebookOptions: StrategyOption = {
+  clientID: oauth2Providers['facebook-login'].clientID,
+  clientSecret: oauth2Providers['facebook-login'].clientSecret,
+  callbackURL: oauth2Providers['facebook-login'].callbackURL,
+  profileFields: oauth2Providers['facebook-login'].profileFields
+};
+
+const googleOptions: GoogleOptions = {
+  clientID: oauth2Providers['google-login'].clientID,
+  clientSecret: oauth2Providers['google-login'].clientSecret,
+  callbackURL: oauth2Providers['google-login'].callbackURL,
+  scope: ['email']
+};
+
+@bind(asAuthStrategy)
+/**
+ * create a common oauth2 strategy for all providers
+ * 
+ * Providers supported: FaceBook
+ *                Todo: google, twitter
+ */
+export class Oauth2Authorization implements AuthenticationStrategy {
+  name = 'Oauth2';
+  facebookStrategy: Strategy;
+  facebookAuthentication: StrategyAdapter<Profile>;
+  googleStrategy: GoogleStrategy;
+  googleAuthentication: StrategyAdapter<Profile>;
+
+  /**
+   * create a common oauth2 authentication object
+   * @param userService 
+   */
+  constructor(
+    @inject(UserServiceBindings.USER_SERVICE)
+    public userService: UserService,
+  ) {
+    this.facebookStrategy = new Strategy(facebookOptions, this.verify.bind(this));
+    this.facebookAuthentication = new StrategyAdapter(
+      this.facebookStrategy,
+      'facebook',
+      mapFaceBookProfile,
+    );
+    this.googleStrategy = new GoogleStrategy(googleOptions, this.verify.bind(this));
+    this.googleAuthentication = new StrategyAdapter(this.googleStrategy, 'google', mapFaceBookProfile);
```

Comment:
We should consider to use `@extensions` to inject different strategies.

Reply:
agreed, @raymondfeng I have added extensions for provider specific implementation now